### ISSUE TITLE
Visual Studio 2015 build is broken

### ISSUE THREAD

**Issue by 1-Olivier** [imageworks/OpenShadingLanguage#1120]
I tried building the latest master on windows and got this:

`..\src\include\OSL/oslnoise.h(228): error C3520: 'ListT': parameter pack must be expanded in this context`

Reverting to just before commit e26cee88efff8096773a54b7052a74d35e88b01c removes the problem.

**Comment by lgritz** [comment]
We're working towards finding a way to reproduce/test this, but our current suspicion is that something about this construct (which all the other compilers agree is valid C++11) is tickling a limitation of VS2015.

Do you know exactly which version of MSVS2015 you are using? Is it the latest? I'm wondering if this is something that was already fixed, if you're a few patches behind, before we have to hunt for a workaround.

**Comment by 1-Olivier** [comment]
It is a fairly old installation of the compiler. cl /? says 19.00.24125.1 which is almost certainly not the latest. You're saying it does build with Visual Studio 2015 on your end?

I will have access to a more recent installation in a day or two. I can check to see if that helps and if it does, upgrading the other compiler should hopefully not be too much trouble.

Also, VS 2015 is not quite fully C++11 compliant, according to this: https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=vs-2015  But I have not looked at the code closely enough to see if the problem is related.

**Comment by lgritz** [comment]
I'm saying I don't have Visual Studio or even Windows at all on my end. The developer who wrote the patch that produces the break on your end did test on MSVS, but he has 2019. At the moment, we don't have very good CI for Windows, its something we need to fix, it would be nice to comprehensively test a few versions, but we just don't right now.

Anyway, before digging in further to try to track this probem down, we just wanted to be sure it wasn't something trivially solved by your using a patched version of VS2015 if one is available.

**Comment by lgritz** [comment]
I'll also add that the current supported release is OSL 1.10, which does not include this latest change and therefore should work fine for you.

The "master" branch is in progress, will some day be a 2.0 release, and I'm not sure we are able to guarantee that MSVS 2015 will be supported by the eventual release. I'm not saying that it won't be, either. Just that it's uncertain because we don't have a release date scheduled and we do from time to time raise the minimum compiler requirements.

**Comment by 1-Olivier** [comment]
Of course, I understand the requirements may change and this code could well be the cause.

Just put this on hold for now. I will get back to you once I've had the chance to try another compiler revision or two. Then you can figure out if you want to update the code or the compiler requirements.

**Comment by lgritz** [comment]
OK, let us know how it goes.

If there is a subsequent patch to MSVS 2015 that makes this work, then I think the expedient thing is to just recommend that the few people on the broken version simply upgrade on their end, rather than our noodling with the code to try to work around a long-ago-fixed MSVS bug.

If NO version of MSVS2015 works with this code, then I think we'll want to try a little bit of code modification, just to see if some small change of equivalent code will work around the limitation. But we'll hold on to the possibility that it's just not fixable and have to assess whether to put more work into it, or just raise the minimum compiler to 2017, which would certainly require a mail list discussion to see if there are lots of people who will still be dependent on MSVS2015 even through the year 2020.

**Comment by 1-Olivier** [comment]
I couldn't actually find a newer version of the VC++ 2015 compiler. I'm not certain one exists. The [wikipedia list](https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering) strongly suggests that unlike 2017 and 2019, 2015 only ever had a single compiler version.

In any case, I tried the VS 2017 tooling (compiler 19.16.27034) and it compiles fine with that. I will let you work out what to do about it, that code has too many templates for me to look at it any longer ;-) I will start preparing for another round of toolchain updates instead.

**Comment by shdwdln** [comment]
@1-Olivier Can you share your build chain? I am getting everything building with vcpkg and it says it can't find LLVM root etc even though I am setting it.
Any thoughts anyone?

**Comment by 1-Olivier** [comment]
@shdwdln I use cmake and set LLVM_DIRECTORY and LLVM_VERSION. Never used vcpkg so I can't help there, you should probably try the mailing list.

I have upgraded to VS2019 now so this issue isn't really one for us anymore.
